In [1]:
import numpy as np
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

In [3]:
df = pd.read_excel('Online_retail.xlsx')
df

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                                burgers,meatballs,eggs                                                                                                                                                                             
1                                               chutney                                                                                                                                                                             
2                                        turkey,avocado                                                                                                                                                                             
3     mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                        low fat yogurt                                                                                                                                                                             
...                                                 ...                                                                                                                                                                             
7495                      butter,light mayo,fresh bread                                                                                                                                                                             
7496  burgers,frozen vegetables,eggs,french fries,ma...                                                                                                                                                                             
7497                                            chicken                                                                                                                                                                             
7498                                 escalope,green tea                                                                                                                                                                             
7499    eggs,frozen smoothie,yogurt cake,low fat yogurt                                                                                                                                                                             

[7500 rows x 1 columns]

In [4]:
df.columns = ['Items']

In [5]:
df

Items
0                                burgers,meatballs,eggs
1                                               chutney
2                                        turkey,avocado
3     mineral water,milk,energy bar,whole wheat rice...
4                                        low fat yogurt
...                                                 ...
7495                      butter,light mayo,fresh bread
7496  burgers,frozen vegetables,eggs,french fries,ma...
7497                                            chicken
7498                                 escalope,green tea
7499    eggs,frozen smoothie,yogurt cake,low fat yogurt

[7500 rows x 1 columns]

In [6]:
df.isna().sum()

Items    0
dtype: int64

In [7]:
X = df['Items'].str.get_dummies(sep=',')

In [8]:
X

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0              0        0                  0          0        0            0   
1              0        0                  0          0        0            0   
2              0        0                  0          0        1            0   
3              0        0                  0          0        0            0   
4              0        0                  0          0        0            0   
...          ...      ...                ...        ...      ...          ...   
7495           0        0                  0          0        0            0   
7496           0        0                  0          0        0            0   
7497           0        0                  0          0        0            0   
7498           0        0                  0          0        0            0   
7499           0        0                  0          0        0            0   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0         0               0          0            0  ...       0   
1         0               0          0            0  ...       0   
2         0               0          0            0  ...       1   
3         0               0          0            0  ...       0   
4         0               0          0            0  ...       0   
...     ...             ...        ...          ...  ...     ...   
7495      0               0          0            0  ...       0   
7496      0               0          0            0  ...       0   
7497      0               0          0            0  ...       0   
7498      0               0          0            0  ...       0   
7499      0               0          0            0  ...       0   

      vegetables mix  water spray  white wine  whole weat flour  \
0                  0            0           0                 0   
1                  0            0           0                 0   
2                  0            0           0                 0   
3                  0            0           0                 0   
4                  0            0           0                 0   
...              ...          ...         ...               ...   
7495               0            0           0                 0   
7496               0            0           0                 0   
7497               0            0           0                 0   
7498               0            0           0                 0   
7499               0            0           0                 0   

      whole wheat pasta  whole wheat rice  yams  yogurt cake  zucchini  
0                     0                 0     0            0         0  
1                     0                 0     0            0         0  
2                     0                 0     0            0         0  
3                     0                 1     0            0         0  
4                     0                 0     0            0         0  
...                 ...               ...   ...          ...       ...  
7495                  0                 0     0            0         0  
7496                  0                 0     0            0         0  
7497                  0                 0     0            0         0  
7498                  0                 0     0            0         0  
7499                  0                 0     0            1         0  

[7500 rows x 120 columns]

In [9]:
X = df['Items'].str.replace(' ', '').str.get_dummies(sep=',')

In [10]:
X

almonds  antioxydantjuice  asparagus  avocado  babiesfood  bacon  \
0           0                 0          0        0           0      0   
1           0                 0          0        0           0      0   
2           0                 0          0        1           0      0   
3           0                 0          0        0           0      0   
4           0                 0          0        0           0      0   
...       ...               ...        ...      ...         ...    ...   
7495        0                 0          0        0           0      0   
7496        0                 0          0        0           0      0   
7497        0                 0          0        0           0      0   
7498        0                 0          0        0           0      0   
7499        0                 0          0        0           0      0   

      barbecuesauce  blacktea  blueberries  bodyspray  ...  turkey  \
0                 0         0            0          0  ...       0   
1                 0         0            0          0  ...       0   
2                 0         0            0          0  ...       1   
3                 0         0            0          0  ...       0   
4                 0         0            0          0  ...       0   
...             ...       ...          ...        ...  ...     ...   
7495              0         0            0          0  ...       0   
7496              0         0            0          0  ...       0   
7497              0         0            0          0  ...       0   
7498              0         0            0          0  ...       0   
7499              0         0            0          0  ...       0   

      vegetablesmix  waterspray  whitewine  wholeweatflour  wholewheatpasta  \
0                 0           0          0               0                0   
1                 0           0          0               0                0   
2                 0           0          0               0                0   
3                 0           0          0               0                0   
4                 0           0          0               0                0   
...             ...         ...        ...             ...              ...   
7495              0           0          0               0                0   
7496              0           0          0               0                0   
7497              0           0          0               0                0   
7498              0           0          0               0                0   
7499              0           0          0               0                0   

      wholewheatrice  yams  yogurtcake  zucchini  
0                  0     0           0         0  
1                  0     0           0         0  
2                  0     0           0         0  
3                  1     0           0         0  
4                  0     0           0         0  
...              ...   ...         ...       ...  
7495               0     0           0         0  
7496               0     0           0         0  
7497               0     0           0         0  
7498               0     0           0         0  
7499               0     0           1         0  

[7500 rows x 119 columns]

In [11]:
frequent_itemsets = apriori(X, min_support=0.01, use_colnames=True)

/Users/anurag_warthi/anaconda3/envs/new_environment/lib/python3.12/site-packages/mlxtend/frequent_patterns/fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


In [12]:
frequent_itemsets

support                               itemsets
0    0.020267                              (almonds)
1    0.033200                              (avocado)
2    0.010800                        (barbecuesauce)
3    0.014267                             (blacktea)
4    0.011467                            (bodyspray)
..        ...                                    ...
254  0.011067       (groundbeef, milk, mineralwater)
255  0.017067  (groundbeef, spaghetti, mineralwater)
256  0.015733        (milk, spaghetti, mineralwater)
257  0.010267    (spaghetti, mineralwater, oliveoil)
258  0.011467    (pancakes, spaghetti, mineralwater)

[259 rows x 2 columns]

In [13]:
rules = association_rules(frequent_itemsets, metric='lift', min_threshold=1.2)
rules.sort_values('lift', ascending=False).head(10)

antecedents                consequents  antecedent support  \
160              (herb&pepper)               (groundbeef)            0.049467   
161               (groundbeef)              (herb&pepper)            0.098267   
329               (groundbeef)  (spaghetti, mineralwater)            0.098267   
328  (spaghetti, mineralwater)               (groundbeef)            0.059733   
343                 (oliveoil)  (spaghetti, mineralwater)            0.065733   
338  (spaghetti, mineralwater)                 (oliveoil)            0.059733   
145                 (tomatoes)         (frozenvegetables)            0.068400   
144         (frozenvegetables)                 (tomatoes)            0.095333   
141                   (shrimp)         (frozenvegetables)            0.071333   
140         (frozenvegetables)                   (shrimp)            0.095333   

     consequent support   support  confidence      lift  representativity  \
160            0.098267  0.016000    0.323450  3.291555               1.0   
161            0.049467  0.016000    0.162822  3.291555               1.0   
329            0.059733  0.017067    0.173677  2.907540               1.0   
328            0.098267  0.017067    0.285714  2.907540               1.0   
343            0.059733  0.010267    0.156187  2.614731               1.0   
338            0.065733  0.010267    0.171875  2.614731               1.0   
145            0.095333  0.016133    0.235867  2.474134               1.0   
144            0.068400  0.016133    0.169231  2.474134               1.0   
141            0.095333  0.016667    0.233645  2.450820               1.0   
140            0.071333  0.016667    0.174825  2.450820               1.0   

     leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
160  0.011139    1.332841       0.732423  0.121457   0.249723    0.243136  
161  0.011139    1.135402       0.772060  0.121457   0.119255    0.243136  
329  0.011197    1.137893       0.727562  0.121097   0.121182    0.229696  
328  0.011197    1.262427       0.697745  0.121097   0.207875    0.229696  
343  0.006340    1.114306       0.661001  0.089120   0.102581    0.164031  
338  0.006340    1.128171       0.656783  0.089120   0.113610    0.164031  
145  0.009613    1.183913       0.639564  0.109304   0.155344    0.202549  
144  0.009613    1.121370       0.658605  0.109304   0.108234    0.202549  
141  0.009866    1.180480       0.637444  0.111111   0.152887    0.204235  
140  0.009866    1.125418       0.654355  0.111111   0.111441    0.204235

- There is a 32.4% probability that when a customer purchases herb&pepper, they also purchase - groundbeef.
- There is a 28.6% probability that when customer purchases mineral water & spaghetti, they also - purchase groundbeef.
- There is a 17.2% probability that when customer purchases mineral water & spaghetti, they also - purchase olive oil.
- There is a 23.6% probability that when customer purchases tomatoes, they also purchase frozen - vegetables.
- There is a 23.4% probability that when customer purchases shrimp, they also purchase frozen - vegetables.

##### Based on the provided table, it is evident that Herb & Pepper, ground beef, mineral water, spaghetti, and olive oil should be stored in a contiguous manner or in a similar row to facilitate customer access and convenience.

##### Tomatoes, frozen vegetables, and shrimp should be grouped together as customers are more inclined to purchase all three items.

## Interview Questions

##### 1. What is lift and why is it important in Association Rules?
Lift measures how much more likely two items are to be bought together compared to if they were bought independently. It helps determine whether the association is meaningful or just due to chance. Lift helps in filtering out trivial or misleading rules by showing real dependency.

##### 2. What are Support and Confidence? How do you calculate them?

Support is how often an item occurs in the dataset. 

Support(A) = (Number of transactions containing A) / (total transactions)

    - Higher support means the item is frequently brought.
    - Low support means the item is rare.

Confidence is how often B appears when A is brought. 

Confidence (A -> B): Support( A U B) / Support(A)

    - High confidence means B is frequently bought when A is bought.

##### 3. What are some limitations or challenges of Association rules mining?
1. Too Many Rules
    - Large datasets generate thousands of rules, many of which are irrelevant.
2. Computational Complexity
    - Apriori needs to scan the dataset multiple times, making it slow for large datasets.
3. Low Support Items are Ignored
    - Rare but highly valuable associations might be ignored due to low support.
4. Misleading Confidence
    - High confidence does not always mean a strong relationship.
5. Handling Categorical & Continuous Data
    - Works best for categorical data